Framework/Algorithms/src/CalculateFlatBackground.cpp

```diff
@@ -152,182 +153,127 @@ void CalculateFlatBackground::exec() {
         break;
       case Modes::MEAN:
-        background = Mean(outputWS, currentSpec, startX, endX);
+        Mean(histogram, background, variance, startX, endX);
         break;
       case Modes::MOVING_AVERAGE:
-        background = MovingAverage(outputWS, currentSpec, windowWidth);
+        MovingAverage(histogram, background, variance, windowWidth);
         break;
       }
-      if (background < 0) {
-        g_log.warning() << "Problem with calculating the background number of "
-                           "counts spectrum with index " << currentSpec
-                        << ". The spectrum has been left unchanged.\n";
-        g_log.debug() << "The background for spectra index " << currentSpec
-                      << "was calculated to be " << background << '\n';
-        continue;
-      } else { // only used for the logging that gets done at the end
-        backgroundTotal += background;
+    }
+    if (background < 0) {
+      g_log.debug() << "The background for spectra index " << i
+                    << "was calculated to be " << background << '\n';
+      g_log.warning() << "Problem with calculating the background number of "
+                         "counts spectrum with index " << i << ".";
+      if (removeBackground) {
+        g_log.warning() << " The spectrum has been left unchanged.\n";
+      } else {
+        g_log.warning() << " The output background has been set to zero.\n";
       }
-
-      // Get references to the current spectrum
-      auto &Y = outputWS->mutableY(currentSpec);
-      auto &E = outputWS->mutableE(currentSpec);
-      // Now subtract the background from the data
-      for (int j = 0; j < blocksize; ++j) {
-        if (removeBackground) {
-          Y[j] -= background;
-        } else {
-          Y[j] = background;
-        }
-        // remove negative values
-        if (m_nullifyNegative && Y[j] < 0.0) {
-          Y[j] = 0;
+      background = 0;
+      variance = 0;
+    }
+    backgroundTotal += background;
+    HistogramData::Histogram outHistogram(histogram);
+    auto &ys = outHistogram.mutableY();
+    auto &es = outHistogram.mutableE();
+    if (removeBackground) {
+      for (size_t j = 0; j < ys.size(); ++j) {
+        double val = ys[j] - background;
+        double err = std::sqrt(es[j] * es[j] + variance);
+        if (nullifyNegative && val < 0) {
+          val = 0;
           // The error estimate must go up in this nonideal situation and the
           // value of background is a good estimate for it. However, don't
           // reduce the error if it was already more than that
-          E[j] = E[j] > background ? E[j] : background;
+          err = es[j] > background ? es[j] : background;
+        }
+        ys[j] = val;
+        es[j] = err;
+      }
+    } else {
+      for (size_t j = 0; j < ys.size(); ++j) {
+        const double originalVal = inputWS->y(i)[j];
+        if (nullifyNegative && background > originalVal) {
```

Comment:
Could you add brackets to make the condition clearer: `(nullifyNegative && (background > originalVal))`